### ISSUE TITLE
macOS Parallel "openps" fails with "lpr -r output1.prn", succeeds with shell script 

### ISSUE THREAD

**Issue by emendelson** [joncampbell123/dosbox-x#2632]
I'm working on a project that uses DOSBox-X under macOS to run word-processing applications. Thanks to @Wengier's recent work, I can easily set different programs to run when a word-processor prints to a PCL or PostScript driver. This works perfectly, except that I can't get the `lpr` command to work correctly. Here's the relevant part of my conf file (DOSBox-X is deep inside an AppleScript app, thus the ../../.. strings.)

```
parallel1 = file file:../../../output1.prn timeout:1000 openpcl:"../../../printpcl.sh output1.prn" openps:"/usr/bin/lpr -r output1.prn"
``` 

When printing a PS file, this always produces an error message; and the same error message happens if I use a shell script to run the `lpr -r` command instead of running it directly. If I run the `lpr` command from a macOS terminal in the same folder with the output1.prn file, the file prints correctly - the command only fails when run from DOSBox-X.

I know that the PostScript file is valid and that PS detection works, because when I print PS to a different port that uses a shell script to process the PS output works correctly. For some reason, I can't get `lpr` to work from the conf file.

I'll continue to work on this, but if anyone has any insight, I'll be grateful for any suggestions.

EDIT: I've worked around this by writing a shell script that converts the PS file to a PDF and then prints the PDF with the `lpr` command. But this doesn't solve one remaining problem: I've set up this system to print raw data to a printer when the user prints to lpt2, by using the command `lpr -l -r output2.prn` - but again, this fails when run from the conf file, so it would be good to have a fix. Is it possible that the "openps" and similar options in the conf file won't work with built-in unix-style commands?

EDIT: And now I was able to solve this for printing with `lpr` by writing shell scripts that use the osascript application to run an AppleScript that runs `lpr`. This is very roundabout - the shell script runs AppleScript which runs a shell command, but it works. It would be much easier if the internal command could be run from the conf file, but maybe that's impossible.

**Comment by Wengier** [comment]
@emendelson For the "openpcl" and "openps" options (and for "openwith" option too), you don't need to specify the output file ("output1.prn" in your case), as it is automatically implied from the "file" option - the command specified by "openpcl" and "openps" options will operate on the output file specified by "file" (the output file automatically becomes the last parameter). So you can for example try `openpcl:../../../printpcl.sh` and `openps:"/usr/bin/lpr -r"` instead. You did not mentioned exactly what error message you saw from the lpr command, but I hope the above will solve this. But if not, posting the actual error message will likely help solve the issue. Hope this helps.

**Comment by emendelson** [comment]
@Wengier Thank you! I was able to fix the `lpr` issue simply by removing the filename from the command string, as you suggested. I was using the filename because my shell scripts required the filename in the command string (I realize that I can probably hard-code the filename into the shell script, but I may use the scripts for more than one output, so it seemed best to use the filename as a parameter that the script can process as "$1".)

The lpr command did not display an error message; I only got the DOSBox-X message saying that the print handler did not complete correctly. I couldn't find an option to display the specific message, but maybe there is one?

So, in summary: My shell scripts need the "output1.prn" filename as a parameter, but lpr does not work if the filename is used as a parameter. Someday I'll study the code and try to figure out why it happens. Meanwhile, the problem is solved.

**Comment by Wengier** [comment]
@emendelson Glad to hear that the problem is already solved. There is no particular option just for displaying error messages from lpr, but if DOSBox-X is launched from the macOS Terminal, you may see the output of the lpr command in the Terminal window.

DOSBox-X automatically appends the file name to the command handler, so if you add it manually then there will be a duplicate parameter, thus cause errors. You can just assume the file name is already provided. Hope this helps.
